### PR TITLE
Discord Server Sync: Remove deleted channels/roles

### PR DESCRIPTION
**Pull request by rshipp**
### Description

In addition to adding/updating channels/roles, we should also remove them from our local db when they've been deleted from the Discord server.

### Motivation and Context

Task: https://app.asana.com/0/1201131148207877/1201314155101891/f

### Screenshots

Screen Shot 2022-01-06 at 4 34 54 PM

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] I have added neccessary labels
- [ ] My PR title follows the convention in the guide
- [ ] I have assigned myself to the PR
- [ ] I have requested reviews from team members
- [ ] I have created a task for reviewers on PM tool
- [ ] All new and existing tests passed.

### CODE REVIEW

File: app/services/sync_discord_servers.rb [hunk 1]
```diff
@@ -33,6 +33,16 @@ def sync_server(server)
   def sync_server_channels(server)
     response = Discordrb::API::Server.channels("Bot #{Rails.application.secrets.discord_bot_token}", server.discord_server_id)
     channels = JSON.parse(response.body)
+
+    # Remove deleted channels.
+    deleted_channels = server.channels.map { |c| c.discord_channel_id } - channels.map { |c| c['id'].to_s }
+    puts server.channels.map { |c| c.discord_channel_id }.to_s
+    puts channels.map { |c| c['id'].to_s }.to_s
```
Comment (elenabaurkot) [inline app/services/sync_discord_servers.rb:40]: I don't know if you wanted to keep the puts here or not, but other than that looks good to me 👍

File: app/services/sync_discord_servers.rb [hunk 2]
```diff
@@ -33,6 +33,16 @@ def sync_server(server)
   def sync_server_channels(server)
     response = Discordrb::API::Server.channels("Bot #{Rails.application.secrets.discord_bot_token}", server.discord_server_id)
     channels = JSON.parse(response.body)
+
+    # Remove deleted channels.
+    deleted_channels = server.channels.map { |c| c.discord_channel_id } - channels.map { |c| c['id'].to_s }
```
Comment (sadleb) [inline app/services/sync_discord_servers.rb:38]: Is there any possibility that the `channels.map` is temporarily empty or incorrect causing us to destroy the local ones accidentally? If that did happen, is it fine b/c the next successful sync would add them back? It doesn't actually destroy anything that we need to safeguard, right?

File: app/services/sync_discord_servers.rb [hunk 3]
```diff
@@ -33,6 +33,16 @@ def sync_server(server)
...
+    puts server.channels.map { |c| c.discord_channel_id }.to_s
+    puts channels.map { |c| c['id'].to_s }.to_s
+    Rails.logger.info("Deleting channels: #{deleted_channels}")
+    Honeycomb.add_field('sync_discord_servers.deleted_channels', deleted_channels)
+    DiscordServerChannel.where(discord_channel_id: deleted_channels).destroy_all
```
Comment (sadleb) [inline app/services/sync_discord_servers.rb:43]: Should we wrap this all in a `if deleted_channels.present?` block?

File: app/services/sync_discord_servers.rb [hunk 4]
```diff
@@ -61,6 +71,14 @@ def sync_server_channels(server)
   def sync_server_roles(server)
     response = Discordrb::API::Server.roles("Bot #{Rails.application.secrets.discord_bot_token}", server.discord_server_id)
     roles = JSON.parse(response.body)
+
+    # Remove deleted roles.
+    deleted_roles = server.roles.map { |c| c.discord_role_id } - roles.map { |c| c['id'].to_s }
+    Rails.logger.info("Deleting roles: #{deleted_roles}")
+    Honeycomb.add_field('sync_discord_servers.deleted_roles', deleted_roles)
+    DiscordServerRole.where(discord_role_id: deleted_roles).destroy_all
```
Comment (sadleb) [inline app/services/sync_discord_servers.rb:79]: Ditto on the q's above. 1) anything to safeguard against if there is a temp blip? 2) wrap in `if blah.present?` block?

File: spec/services/sync_discord_servers_spec.rb [hunk 5]
```diff
@@ -97,6 +97,30 @@
         expect(DiscordServerChannel.find_by(name: text_channel[:name])).not_to eq(nil)
       end
     end
+
+    context 'with deleted channel' do
```
Comment (sadleb) [inline spec/services/sync_discord_servers_spec.rb:101]: Are there any tests that cover the case where the local db channels match the server channels? I'm just thinking about making sure this works in that case: `DiscordServerChannel.where(discord_channel_id: deleted_channels).destroy_all`

File: app/services/sync_discord_servers.rb (hunk 1 above)
Comment (rshipp) [reply app/services/sync_discord_servers.rb]: whoops, nope

File: app/services/sync_discord_servers.rb (hunk 2 above)
Comment (rshipp) [reply app/services/sync_discord_servers.rb:38]: anything is possible, but it's not likely. that's like asking if technically, could a salesforce api call to get participants return an empty list. and correct, we wouldn't really care, the discord scheduler would just be broken until the next day or we rerun the task.

File: spec/services/sync_discord_servers_spec.rb (hunk 5 above)
Comment (rshipp) [reply spec/services/sync_discord_servers_spec.rb:101]: there's some directly above this. but `Model.where(something: []).destroy_all` translates to `delete from table where something in ()`, which does nothing.

File: app/services/sync_discord_servers.rb (hunk 3 above)
Comment (rshipp) [reply app/services/sync_discord_servers.rb:41]: why? an extra branch to remove a no-op SQL call doesn't seem worth it

**Review by sadleb — APPROVED** [review]
Thanks for answering my questions. lgtm

**Pull request merged**